### ISSUE TITLE
Feature request: Add "Lint/RaiseNotImplementedError" cop

### ISSUE THREAD

**Issue by johnnyshields** [rubocop/rubocop#13042]
I suggest to have a Cop `raise Lint/RaiseNotImplementedError` which checks for any usage of NotImplementedError and suggests to replace it with `NoMethodError` (unsafe autocorrect).

```ruby
# BAD
def foobar
  raise NotImplementedError.new("Subclass must define foobar")
end

# GOOD
def foobar
  raise NoMethodError.new("Subclass must define foobar")
end
```

The reason is that NotImplementedError does not extend StandardError, so if you follow the `Lint/RescueException` cop and convert to `rescue StandardError` you'll actually be _not catching_ such errors. For more details:
https://oleg0potapov.medium.com/ruby-notimplementederror-dont-use-it-dff1fd7228e5

**Comment by dvandersluis** [comment]
1. How would you prevent false positives for the "intended usage" of `NotImplementedError`? It's not invalid it's just potentially misused.
2. Why would it not extending `StandardError` matter? Why would you want to rescue `NotImplementedError`, especially when it's used for an abstract class. I think @pirj makes a similar argument here: https://github.com/rubocop/ruby-style-guide/issues/458#issuecomment-1174065851

If this is adopted it should be a `Style` cop and disabled by default IMO. But even then I'm not sure if it makes sense -- why should `NotImplementedError` be treated differently than every other built in exception class? @rubocop/rubocop-core  do you have any thoughts?

**Comment by dvandersluis** [comment]
Another reference for this discussion: https://bugs.ruby-lang.org/issues/18915

A patch was proposed that changes the documentation to explicitly mention `NotImplementedError`'s use in abstract methods. Members of ruby core also discussed being ok with `NotImplementedError` being used in such a capacity, as well as discussing potential new exception classes to use instead. Nothing was decided of course, but this is some more information.

**Comment by johnnyshields** [comment]
@dvandersluis The reason is that StandardError is usually for user/app/gem originated exceptions while Exception is includes things like Sigterm which you dont want to be rescuing in your app/user-land code. There is already a cop which prevents you from rescuing Exception.

**Comment by pirj** [comment]
I still think we should not implement the cop as suggested here:
1. NotImplementedError prevails 
2. It will raise early and has less chances to be rescued

 If the Ruby doc patch merges, it’s official, and we will need a cop to flag NoMethodError instead abstract methods instead

**Comment by dvandersluis** [comment]
I agree with @pirj. It is not settled that `NotImplementedError` is incorrect when it's being used, or that `NoMethodError` is correct.

**Comment by johnnyshields** [comment]
@dvandersluis @pirj Irrespective of what Ruby may or may not do in the future, raising a `NotImplementedError` semantically violates the [Lint/RaiseException](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/RaiseException) cop. This is because:
- `NotImplementedError` is **not** a subclass of `StandardError`
- `NotImplementedError` is a subclass of `Exception`

So I think at a minimum, if `Lint/RaiseException` is enabled, we should be requiring `NoMethodError` instead of `NotImplementedError`.

(For the minority of users who are fine to raise `Exception` (not `StandardError`) in their app code, then doing this is of course no issue.)

**Comment by dvandersluis** [comment]
`NotImplementedError` is a subclass of `ScriptError`, not `Exception`. But regardless, `Lint/RaiseException` is specifically for raising `Exception`, and does not register for any subclass of `Exception`.
